src/main.cpp

```diff
@@ -5894,6 +5931,13 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
         int count=0;
         size_t i;
+        if (pfrom->xThinBlockHashes.size() != pfrom->thinBlock.vtx.size())  // Because the next loop assumes this
+          {
+            LogPrint("thin", "Inconsistent thin block data.  Aborting the thin block\n");
+            // TODO clear out the thin block
```

Comment:
Where the TODO is you should insert:

            {
            LOCK(cs_vNodes);
            pfrom->mapThinBlocksInFlight.erase(inv.hash); 
            pfrom->thinBlockWaitingForTxns = -1;
            pfrom->thinBlock.SetNull();
            }

            // Clear the thinblock timer used for preferential download
            thindata.ClearThinBlockTimer(inv.hash);